### PR TITLE
Handle error response for odoo.connect()

### PR DESCRIPTION
**Pull request by martijnthe**
When the underlying request would return an error response (for some reason HTTP status 200 is used... 😞), the `connect()` promise was still being resolved to a `Odoo` object, instead of being rejected.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -25,7 +25,6 @@
     "connect",
     "json",
     "rpc",
-    "connect",
```
Comment (SamVerschueren) [inline package.json:27]: Didn't notice I had a duplicate. Care to create a separate PR for this instead of doing it inside this one? I like to not have unrelated changes.

File: lib/odoo.js [hunk 2]
```diff
@@ -38,11 +38,17 @@ class Odoo {
...
 		return got.post(`${config.host}:${config.port}/web/session/authenticate`, {json: true, headers, body})
-			.then(res => new OdooClient(res.body.result, {
-				host: config.host,
-				port: config.port,
-				sid: res.headers['set-cookie'][0].split(';')[0]
-			}));
+			.then(res => {
+				if (res.body.error) {
+					// HTTP Status is 200 / OK!
+					throw new Error(res.body.error.message);
```
Comment (SamVerschueren) [inline lib/odoo.js]: `throw res.body.error`

File: lib/odoo.js [hunk 3]
```diff
@@ -38,11 +38,17 @@ class Odoo {
...
+				}
```
Comment (SamVerschueren) [inline lib/odoo.js:52]: Add a newline after the if-statement.

File: test/fixtures/setup.js [hunk 4]
```diff
@@ -56,4 +54,36 @@ module.exports = test => {
 				}
 			});
 	});
+
+	test.before(() => {
```
Comment (SamVerschueren) [inline test/fixtures/setup.js]: Just put the code inside the other `test.before` so you don't have 2 `test.before` blocks.

File: test/fixtures/setup.js [hunk 5]
```diff
@@ -56,4 +54,36 @@ module.exports = test => {
...
+				json: true,
+				headers,
+				body: JSON.stringify(failBody)
+			})
+			.resolves({
+				body: {
+					jsonrpc: '2.0',
+					id: null,
+					error: {
+						message: "Odoo Server Error",
```
Comment (SamVerschueren) [inline test/fixtures/setup.js]: Single quotes

File: test/fixtures/setup.js [hunk 6]
```diff
@@ -56,4 +54,36 @@ module.exports = test => {
...
+						code: 200,
+						data: {
+							debug: "Traceback ..."
```
Comment (SamVerschueren) [inline test/fixtures/setup.js]: Single quotes

File: test/test.js [hunk 7]
```diff
@@ -33,3 +33,8 @@ test('connect', async t => {
 		sid: 'session_id=2ad18d37785fab31c8e9e3cb8c21f8917f48b63b'
 	});
 });
+
+test('connect-fail', async t => {
```
Comment (SamVerschueren) [inline test/test.js]: `test('connect failure', async t => {`

File: test/test.js [hunk 8]
```diff
@@ -33,3 +33,8 @@ test('connect', async t => {
...
+	const odoo = new Odoo({host: 'foobar.com'});
+	t.throws(odoo.connect({database: 'fail', username: 'fail', password: 'fail'}));
```
Comment (SamVerschueren) [inline test/test.js]: `t.throws(odoo.connect({database: 'fail', username: 'fail', password: 'fail'}), 'Odoo Server Error');`

**Review by SamVerschueren — CHANGES_REQUESTED** [review]
Thank you very much for the PR! Good quality! Added some comments that should be addressed first.

File: lib/odoo.js (hunk 2 above)
Comment (martijnthe) [reply lib/odoo.js]: It looks like Ava's `throws()` doesn't like a `Promise` that is rejected with a `string`:
https://github.com/avajs/ava/blob/master/test/promise.js#L213

Personally I like to use `Error` because it captures the stack trace at the moment it is created.

What's you motivation for wanting `string` here?

Comment (martijnthe) [reply lib/odoo.js]: I guess I can work-around the limitation of `throws()` by doing:

```
test('connect failure', async t => {
	const odoo = new Odoo({host: 'foobar.com'});
	const error = await t.throws(odoo.connect({database: 'fail', username: 'fail', password: 'fail'}));
	t.is(error, 'Odoo Server Error');
});
```

**Comment by martijnthe** [comment]
@SamVerschueren I updated the PR. Please take a look again. Thanks!

File: lib/odoo.js (hunk 2 above)
Comment (SamVerschueren) [reply lib/odoo.js]: I agree, you should always reject with an error. But isn't `res.body.error` an Error object? It's definitely not a string.

**Comment by martijnthe** [comment]
It is actually a string. It's part of the body response JSON.

On Jan 8, 2017 10:36 PM, "Sam Verschueren" <notifications@github.com> wrote:

*@SamVerschueren* commented on this pull request.
------------------------------

In lib/odoo.js <https://github.com/pridiktiv/odoo-connect/pull/3>:

> @@ -38,11 +38,17 @@ class Odoo {
 		};

 		return got.post(`${config.host}:${config.port}/web/session/authenticate`,
{json: true, headers, body})
-			.then(res => new OdooClient(res.body.result, {
-				host: config.host,
-				port: config.port,
-				sid: res.headers['set-cookie'][0].split(';')[0]
-			}));
+			.then(res => {
+				if (res.body.error) {
+					// HTTP Status is 200 / OK!
+					throw new Error(res.body.error.message);

I agree, you should always reject with an error. But isn't res.body.error
an Error object? It's definitely not a string.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/pridiktiv/odoo-connect/pull/3>, or mute the thread

File: lib/odoo.js [hunk 9]
```diff
@@ -38,11 +38,18 @@ class Odoo {
...
 		return got.post(`${config.host}:${config.port}/web/session/authenticate`, {json: true, headers, body})
-			.then(res => new OdooClient(res.body.result, {
-				host: config.host,
-				port: config.port,
-				sid: res.headers['set-cookie'][0].split(';')[0]
-			}));
+			.then(res => {
+				if (res.body.error) {
+					// HTTP Status is 200 / OK!
+					throw res.body.error.message;
```
Comment (SamVerschueren) [inline lib/odoo.js]: I'm saying that you should throw `res.body.error`, not `res.body.error.message`. Thus, `res.body.error` is an object

```js
{
    message: 'Odoo Server Error',
    code: 200,
    data: {
        debug: 'Traceback ...'
    }
}
```

**Comment by martijnthe** [comment]
Oh I see. Yeah,  but still not an Error object.

By coincidence the res.body.error object has a property called "message",
just like an Error object. It feels more brittle to reject with that object
vs throwing an actual Error object, because it depends on the Odoo server
whether "message" is part of it. On top of that, creating an actual Error
would capture the current backtrace, which is nice for debugging.

It's your call...


On Jan 9, 2017 9:08 AM, "Sam Verschueren" <notifications@github.com> wrote:

*@SamVerschueren* commented on this pull request.
------------------------------

In lib/odoo.js
<https://github.com/pridiktiv/odoo-connect/pull/3#pullrequestreview-15656061>
:

> @@ -38,11 +38,18 @@ class Odoo {
 		};

 		return got.post(`${config.host}:${config.port}/web/session/authenticate`,
{json: true, headers, body})
-			.then(res => new OdooClient(res.body.result, {
-				host: config.host,
-				port: config.port,
-				sid: res.headers['set-cookie'][0].split(';')[0]
-			}));
+			.then(res => {
+				if (res.body.error) {
+					// HTTP Status is 200 / OK!
+					throw res.body.error.message;

I'm saying that you should throw res.body.error, not res.body.error.message.
Thus, res.body.error is an object

{
    message: 'Odoo Server Error',
    code: 200,
    data: {
        debug: 'Traceback ...'
    }
}

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/pridiktiv/odoo-connect/pull/3#pullrequestreview-15656061>,
or mute the thread

**Comment by SamVerschueren** [comment]
@martijnthe Let's create an error object out of that one by not only copying the message.

```js
const err = res.body.error;

const error = new Error(err.message);
error.code = err.code;
error.data = err.data;
```

And then throw that `error` object. I just don't want to loose that data cause it might be important. Maybe wrap that in a `createError` function.

```js
throw createError(res.body.error);
```

**Comment by martijnthe** [comment]
SGTM!

I updated the PR.

**Comment by SamVerschueren** [comment]
Can you rebase from `master` again? I solved the linting issues.

**Comment by martijnthe** [comment]
@SamVerschueren rebased

**Comment by SamVerschueren** [comment]
Thanks for this PR @martijnthe! :beers:
